docs/design/features/byreflike-generics.md

```diff
@@ -114,21 +114,103 @@ Adding `gpAcceptByRefLike` to the metadata of a Generic parameter will be consid
 
 Enumerating of constructors/methods on `Span<T>` and `ReadOnlySpan<T>` may throw `TypeLoadException` if `T` is a ByRefLike type. See "Troublesome APIs" above for the list of APIs that cause this condition.
 
-## <a name="special_il_sequences"></a> Special IL Sequences
+## <a name="invalid_il_options"></a> Options for invalid IL
 
-The following are IL sequences involving the `box` instruction. They are used for common C# language constructs and shall continue to be valid, even with ByRefLike types, in cases where the result can be computed at JIT compile time or interpretation and elided safely. These sequences **must** now be elided when the target type is ByRefLike. The conditions where each sequence is elided are described below and each condition will be added to the ECMA-335 addendum.
+There are two potential options below for how to address this issue.
 
-`box` ; `unbox.any` &ndash; Becomes a NOP, if the box target type is equal to the unboxed target type.
+The first indented IL sequences below represents the `is-type` sequence. Combining the first with the second indented section represents the "type pattern matching" scenario in C#. The below sequence performs a type check and then, if successful, consumes the unboxed instance.
 
-`box` ; `br_true/false` &ndash; Becomes the constant `true`, if the box target type is non-`Nullable<T>`.
+```IL
+// Type check
+ldarg.0
+    box <Source>
+    isinst <Target_A>
+    brfalse.s NOT_INST
 
-`box` ; `isinst` ; `unbox.any` &ndash; Becomes a NOP, if the box, `isinst`, and unbox target types are all equal.
+// Unbox and store unboxed instance
+ldarg.0
+    box <Source>
+    isinst <Target_A>
+    unbox.any <Target_A>
+stloc.X
 
-`box` ; `isinst` ; `br_true/false` &ndash; Can become a constant, if the box target type is ByRefLike or the box target type is `Nullable<T>` and target type equalities are computed to be equal. This sequence can be interpreted as behaving as if the ByRefLike nature of the input type doesn't exist.
+NOT_INST:
+ret
+```
+
+With the above IL composition implemented, the following C# describes the following "type pattern matching" scenarios and what one might expect given current C# semantics.
+
+```csharp
+struct S {}
+ref struct RS {}
+interface I {}
+class C {}
+class C<T> {}
+
+// Not currently valid C#
+void M<T, U>(T t) where T: allows ref struct
+{
+    if (t is int i)        // Valid
+    if (t is S vc)         // Valid
+    if (t is RS rs)        // Valid
+    if (t is Span<char> s) // Valid
+    if (t is string str)   // Valid
+    if (t is I itf)        // Valid
+    if (t is C c)          // Valid
+    if (t is C<I> ci)      // Valid
+    if (t is C<U> cu)      // Invalid
+    if (t is U u)          // Invalid
+    if (t is Span<U> su)   // Invalid
+}
+```
```

Comment:
Can you help us understand a bit how this behaves with actual types. Consider as an example: 

```csharp
// Not real C#, just using as an example to understand the IL semantics
ref struct RS2 : I { }

// Which of the `if` succeed and which fail here?
M<RS2, I>(); 
```

I'm particularly interested in the following ones:

```csharp
    if (t is I itf) 
    if (t is object)   // I added this one
    if (t is U u)   
```

Sorry if this seems odd but knowing how IL would execute these helps us understand how the C# rules need to be constructed. Consider concretely how C# works today with non-generic code

```csharp
ref struct S : I { }
interface I { }
void Example(S s) {
  // Warning: expression always false
  if (s is I) 
}

// This is what the code would look like if the user decided to make it generic. 
void GenericExample<T>(T t ) where T: allows ref struct {
  // What do we do here?
  if (t is I) 
}
```

Knowing how that will execute in IL really helps us nail down what we want to recommend to C# LDM on what the C# rules will be. Think we already have a fairly good understanding of what we want to recommend but really want to _know_ the IL semantics to ensure the design is right.

Reply:
Given the call `M<RS2, I>()`:

```csharp
    if (t is I itf) 
```
Can be made to work, but the type local would need to be `RS2` since it cannot be boxed, so perhaps this is "Invalid" from the C# perspective. The check is safe to make (that is, `if (t is I)`).

```csharp
    if (t is object)
```
The `object` check will evaluate to `true` in this PR. It is unlikely we would change that behavior if we adopted the IL instruction approach. We could update the helper functions to special case ByRefLike behavior or define in a special manner to be used when the `allows ref struct` constraint is added.

```csharp
    if (t is U u)
```
This will practically always result in a `false` condition for obvious reasons. If `U` were constrained to something it could potential be the same as `if (t is I itf)`.

Going down the IL instruction approach we need more precision, but adopting the Compiler helpers means we can permit more. The helpers could be "general purpose" in the sense they entirely replace the aforementioned sequences or we could make the helpers ByRefLike only and for example evaluate `t is object` to `false`.

I don't think anyone on the runtime team has a strong opinion about the Compiler helpers route, but if we continue down the IL sequence path, I think we need to know C# wants and then we can say what is possible to guarantee.

```csharp
// This is what the code would look like if the user decided to make it generic. 
void GenericExample<T>(T t ) where T: allows ref struct {
  // What do we do here?
  if (t is I) 
}
```

The "check", `t is I`, is always possible based on this PR and the proposed Compiler helper would provide the same behavior at present. The question though becomes, what would one want to do in that branch if it evaluated to `true`?

Reply:
> The object check will evaluate to true in this PR.
> The "check", t is I, is always possible based on this PR and the proposed Compiler helper would provide the same behavior at present

I understand why this is sensible from an IL / CLR perspective. From a C# perspective though it's not in alignment with how ref safety and type checks work. Both of these types of checks evaluated to `false` at compile time with concrete types + a warning. Given this behavior from IL, in generic contexts we'd need to block type checks with potentially fall into these categories with errors. That is the only way to ensure there is a consistent experience between generic and non-generic code. 

That is not necessarily wrong for runtime or C#. It's nice when the C# and IL type system are in full agreement. But there are other cases where aren't . 

Note: in retrospect kinda wish we'd pushed for an error in the non-generic case.

Reply:
If we go with helper methods, we can make them do whatever works best for C# in these corner cases.

Reply:
> If we go with helper methods, we can make them do whatever works best for C# in these corner cases.

Agreed. This concept was what made Jan's statement so appealing. We just need to define what makes sense for C# and then the really hard part, naming the APIs...

Reply:
`bool ByRefLikeAwareIsInstanceOf<TFrom, TTo>(TFrom source)` and `TTo ByRefLikeAwareCastTo<TFrom, TTo>(TFrom source)`?